src/Types/TokenName.purs

```diff
@@ -0,0 +1,60 @@
+module Types.TokenName
+  ( TokenName
+  , adaToken
+  , getTokenName
+  , mkTokenName
+  , mkTokenNames
+  , tokenNameFromAssetName
+  , assetNameName
+  ) where
+
+import Prelude
+
+import Data.BigInt (BigInt)
+import Data.Bitraversable (ltraverse)
+import Data.Map (Map)
+import Data.Map (fromFoldable) as Map
+import Data.Maybe (Maybe(Nothing))
+import Data.Traversable (class Traversable, traverse)
+import Data.Tuple.Nested (type (/\))
+import FromData (class FromData)
+import Serialization.Types (AssetName) as CSL
+import ToData (class ToData)
+import Types.ByteArray (ByteArray, byteLength)
+
+newtype TokenName = TokenName ByteArray
+
+derive newtype instance Eq TokenName
+derive newtype instance FromData TokenName
+derive newtype instance Ord TokenName
+derive newtype instance ToData TokenName
+
+instance Show TokenName where
+  show (TokenName tn) = "(TokenName " <> show tn <> ")"
+
+getTokenName :: TokenName -> ByteArray
+getTokenName (TokenName tokenName) = tokenName
+
+-- | The empty token name.
+adaToken :: TokenName
+adaToken = TokenName mempty
+
+-- | Create a `TokenName` from a `ByteArray` since TokenName data constructor is
+-- | not exported
+mkTokenName :: ByteArray -> Maybe TokenName
+mkTokenName byteArr =
+  if byteLength byteArr <= 32 then pure $ TokenName byteArr else Nothing
```

Comment:
Is this raw bytes or CBOR encoded? (in terms of CSL, is it `AssetName.new` or `AssetName.from_bytes`?) 

I believe CBOR could contribute extra bytes to the length, we should be extra careful here and be sure it's not CBOR.

Reply:
It should be raw bytes. At least that's how it's used elsewhere in the codebase. Once we have #234 in place this will be clearer